### PR TITLE
Add holes for expressions in function clauses and inference support 

### PR DESCRIPTION
**Pull request by janmasrovira**
This PR adds support for type holes that can be automatically inferred.
Currently, holes can only appear in expressions (function bodies).

For instance:
Before we had:
```
map : (A : Type) → (B : Type) →
  (A → B) → List A → List B;
map A B f nil ≔ nil B ;
map A B f (cons x xs) ≔ cons B (f x) (map A B f xs);
```
and now:
```
map : (A : Type) → (B : Type) →
  (A → B) → List A → List B;
map _ _ f nil ≔ nil _;
map _ _ f (cons x xs) ≔ cons _ (f x) (map _ _ f xs);
```

Proper error messages still need to be added

### CODE REVIEW

File: src/MiniJuvix/Syntax/Abstract/Language/Extra.hs [hunk 1]
```diff
@@ -26,12 +26,9 @@ smallerPatternVariables = \case
 
 viewApp :: Expression -> (Expression, [Expression])
 viewApp e = case e of
-  ExpressionIden {} -> (e, [])
   ExpressionApplication (Application l r) ->
     second (`snoc` r) (viewApp l)
-  ExpressionUniverse {} -> (e, [])
-  ExpressionFunction {} -> (e, [])
-  ExpressionLiteral {} -> (e, [])
+  _ -> (e, [])
```
Comment (jonaprieto) [inline src/MiniJuvix/Syntax/Abstract/Language/Extra.hs:31]: Revert this change. We list all the patterns.

**Comment by jonaprieto** [comment]
Good Job 🏁

File: src/MiniJuvix/Syntax/Abstract/Language/Extra.hs (hunk 1 above)
Comment (janmasrovira) [reply src/MiniJuvix/Syntax/Abstract/Language/Extra.hs:31]: I agree that the rule of thumb should be to always list all the patterns, but in this case we are unfolding an application. There is only one constructor that we need to consider. Moreover, even if we add more constructors to Expression, the only relevant constructor will remain Application
